PowerShellTools/ServiceManagement/PowershellHostProcessFactory.cs

```diff
@@ -0,0 +1,139 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using PowerShellTools.Common;
+
+namespace PowerShellTools.ServiceManagement
+{
+    /// <summary>
+    /// Initilize a process for hosting the WCF service.
+    /// </summary>
+    internal static class PowershellHostProcessFactory
+    {
+        private static Lazy<PowershellHostProcess> _powershellHostProcess;
+        private static object _syncObject = new object();
+
+        static PowershellHostProcessFactory()
+        {
+            LazyCreatePowershellHostProcess();
+        }
+
+        /// <summary>
+        /// The host process we want.
+        /// </summary>
+        internal static PowershellHostProcess HostProcess { get; set; }
+
+        /// <summary>
+        /// In case we need to change host process at some point. What we need to do is signal the termination of the process and then create a new one. 
+        /// </summary>
+        internal static void SignalProcessTerminated()
+        {
+            lock (_syncObject)
+            {
+                LazyCreatePowershellHostProcess();
+            }
+        }
+
+        internal static PowershellHostProcess EnsurePowershellHostProcess()
+        {
+            return _powershellHostProcess.Value;
+        }
+
+        private static PowershellHostProcess CreatePowershellHostProcess()
+        {
+            Process powershellHostProcess = new Process();
+            string hostProcessReadyEventName = Constants.ReadyEventPrefix + Guid.NewGuid();
+            Guid endPointGuid = Guid.NewGuid();
+
+            string exeName = Constants.PowershellHostExeName;
+            string currentPath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
+            string path = Path.Combine(currentPath, exeName);
+            string hostArgs = String.Format(CultureInfo.InvariantCulture,
+                                            "{0}{1} {2}{3} {4}{5}",
+                                            Constants.UniqueEndpointArg, endPointGuid, // For generating a unique endpoint address 
+                                            Constants.VsProcessIdArg, Process.GetCurrentProcess().Id,
+                                            Constants.ReadyEventUniqueNameArg, hostProcessReadyEventName); 
+
+            powershellHostProcess.StartInfo.Arguments = hostArgs;
+            powershellHostProcess.StartInfo.FileName = path;
+            powershellHostProcess.StartInfo.CreateNoWindow = true; 

```

Comment:
Sure. But I’ll leave it as it is now.

From: Eric Deng [mailto:notifications@github.com]
Sent: Wednesday, January 21, 2015 11:34 AM
To: Microsoft/poshtools
Cc: Huanhuan Sun
Subject: Re: [poshtools] Infrastructure for out-of-proc, includes initializing a process and creating a WCF channel for servicing (#36)


In PowerShellTools/ServiceManagement/PowershellHostProcessFactory.cs<https://github.com/Microsoft/poshtools/pull/36#discussion_r23327009>:

> +            Process powershellHostProcess = new Process();

> +            string hostProcessReadyEventName = Constants.ReadyEventPrefix + Guid.NewGuid();

> +            Guid endPointGuid = Guid.NewGuid();

> +

> +            string exeName = Constants.PowershellHostExeName;

> +            string currentPath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);

> +            string path = Path.Combine(currentPath, exeName);

> +            string hostArgs = String.Format(CultureInfo.InvariantCulture,

> +                                            "{0}{1} {2}{3} {4}{5}",

> +                                            Constants.UniqueEndpointArg, endPointGuid, // For generating a unique endpoint address

> +                                            Constants.VsProcessIdArg, Process.GetCurrentProcess().Id,

> +                                            Constants.ReadyEventUniqueNameArg, hostProcessReadyEventName);

> +

> +            powershellHostProcess.StartInfo.Arguments = hostArgs;

> +            powershellHostProcess.StartInfo.FileName = path;

> +            powershellHostProcess.StartInfo.CreateNoWindow = true;

nit: Window could be useful during debugging, can we do #if DEBUG to show window for future convenience?

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/poshtools/pull/36/files#r23327009>.